File: LeetCodeSolutions/LeetCode_0032.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        if not s:
            return 0
        dp = [0] * len(s)
        res = 0
        for _ in range(len(s)):
            if _ > 0:
                if s[_] == '(':
                    dp[_] = 0
                if s[_] == ')':
                    if s[_ - 1] == '(':
                        dp[_] = dp[_ - 2] + 2
                    if s[_ - 1] == ')' and s[_ - dp[_ - 1] - 1] == '(':
                        dp[_] = dp[_ - 1] + 2 + dp[_ - dp[_ - 1] - 2]
                    if dp[_] > res:
                        res = dp[_]
        return res

File: LeetCodeSolutions/LeetCode_0032.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        stack<int> stk;
        stk.push(-1);
        int ret = 0;
        for(int i = 0; i < s.size(); ++i){
            if (s[i] == '(') stk.push(i);
            else{
                stk.pop();
                if(stk.empty()) stk.push(i);
                else ret = max(ret, i - stk.top());
            }
        }
        return ret;
    }
};


class Solution {
public:
    int longestValidParentheses(string s) {
        int n = s.size();
        vector<int> dp(n, 0);
        int ret = 0;
        for(int i = 0; i < n; ++i){
            if (s[i] == ')'){
                if (i > 0 && s[i - 1] == '('){
                    if(i > 1) dp[i] = dp[i - 2] + 2;
                    else dp[i] = 2;
                }
                else{
                    if(i > 1 && dp[i - 1] > 0){
                        int tmp = i - dp[i - 1] - 1;
                        if (tmp < 0) continue;
                        if (tmp == 0 && s[0] == '(') dp[i] = dp[i - 1] + 2;
                        if(tmp > 0  && s[tmp] == '(') dp[i] = 2 + dp[i - 1] + dp[tmp - 1];
                    }
                }
            }
            ret = max(ret, dp[i]);
        }
        return ret;
    }
};